### PR TITLE
Wires refactor

### PR DESCRIPTION
**Pull request by vulppine**
# Wires refactor (again)

Changes wires from a hard-coded layout within components into a prototype, modular layout that allows you to compose each wire.

Each wire can now be added to a *layout* in the prototype type of `wireLayout`, and these actions will be flyweighted across any entity that uses that given wire layout in its component.

Wires themselves are now objects compatible with a new interface, `IWireAction`. This interface handles what happens when a wire is cut, mended, or pulsed. It should also handle what a wire's corresponding status light should look like, if there is one, as well as what happens when the system needs to update a wire's status.

General wire logic is now centralized within `WiresSystem`, and also includes an internal timing update which allows for timed wire action cancellations (i.e., power shorts out for only some time, access shorts out for a limited amount of time).

There are no visual changes otherwise, however, there are certainly feature changes:
- Power wires will now shock players when cut or pulsed without any insulated gloves
- The access wire, once missing, has now been readded to entities that had it in SS13
- The safety door wire now shorts when pulsed, and the timing door wire shorts when pulsed.

## TODO

Reimplement:
- [X] Doors
- [X] Vending machines (the rest of vending machines should also be its own PR)
- [x] Arcades
- [x] Air/fire alarms
- [ ] ~~Particle Accelerator~~ this should be its own PR

Create generic wire types for:
- [x] Power
- [x] Access

Maybe:
- [x] Multiple wire counts for the same action
- [x] Decoy/dummy wires

**Changelog**

:cl:
- tweak: Nanotrasen has finally kicked the old board vendor square in the ass, and told them to actually make their boards more maintainable. Expect more functionality when you pulse, cut, and mend wires in some electronics.

### CODE REVIEW

**Comment by Peptide90** [comment]
Just to add my 2 cents since it was discussed heavily in ideaguys a few times, I see you have multiple wires in the maybe column. Logically not all wires will create an alert when cut however, once we have AI, having multiple wires that alarm is realistic and should discourage hacking.

Otherwise, welcome refactor :)

**Comment by vulppine** [comment]
> Just to add my 2 cents since it was discussed heavily in ideaguys a few times, I see you have multiple wires in the maybe column. Logically not all wires will create an alert when cut however, once we have AI, having multiple wires that alarm is realistic and should discourage hacking.
> 
> Otherwise, welcome refactor :)

yeah, this also makes it so that an implementer can just have repeat wires that all do similar things without having to create duplicate classes (bad) to have similar wires

File: Content.Client/Entry/EntryPoint.cs [hunk 1]
```diff
@@ -77,7 +77,7 @@ public override void Init()
             factory.RegisterClass<SharedLatheComponent>();
             factory.RegisterClass<SharedSpawnPointComponent>();
             factory.RegisterClass<SharedVendingMachineComponent>();
-            factory.RegisterClass<SharedWiresComponent>();
+            // factory.RegisterClass<SharedWiresComponent>();
```
Comment (metalgearsloth) [inline Content.Client/Entry/EntryPoint.cs:80]: real

File: Content.Client/Wires/UI/WiresMenu.cs [hunk 2]
```diff
@@ -14,7 +14,8 @@
 using Robust.Shared.IoC;
 using Robust.Shared.Localization;
 using Robust.Shared.Maths;
-using static Content.Shared.Wires.SharedWiresComponent;
+using Robust.Shared.Random;
+// using static Content.Shared.Wires.SharedWiresComponent;
```
Comment (metalgearsloth) [inline Content.Client/Wires/UI/WiresMenu.cs:18]: real

File: Content.Client/Wires/Visualizers/WiresVisualizer.cs [hunk 3]
```diff
@@ -1,7 +1,8 @@
+using Content.Shared.Wires;
 using Robust.Client.GameObjects;
 using Robust.Shared.GameObjects;
 using Robust.Shared.IoC;
-using static Content.Shared.Wires.SharedWiresComponent;
+// using static Content.Shared.Wires.SharedWiresComponent;
```
Comment (metalgearsloth) [inline Content.Client/Wires/Visualizers/WiresVisualizer.cs:5]: real

**Review by metalgearsloth — APPROVED** [review]
Should seal classes where possible plus some of the old stuff was commented out that needs removing.

File: Content.Client/Wires/UI/WiresMenu.cs [hunk 4]
```diff
@@ -267,7 +268,16 @@ public void Populate(WiresBoundUserInterfaceState state)
...
+            var originalStatuses = new List<StatusEntry>(state.Statuses); // TODO: maybe not this way?
+            var shuffledStatuses = new List<StatusEntry>();
+            for (var i = originalStatuses.Count; i > 0; i--)
+            {
+                var index = random.Next(originalStatuses.Count);
+                shuffledStatuses.Add(originalStatuses[index]);
+                originalStatuses.RemoveAt(index);
+            }
+
+            foreach (var status in shuffledStatuses)
```
[the remark this answers is not in the document]
Comment (vulppine) [reply Content.Client/Wires/UI/WiresMenu.cs:280]: needs to be a seeded random, so it uses the C# random for this

File: Content.Server/Wires/WireLayout.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+// wires. Once one of these is initialized, it should be stored in the
+// WiresSystem as a functional wire set.
+[Prototype("wireLayout")]
+public sealed class WireLayoutPrototype : IPrototype
+{
+    [IdDataFieldAttribute]
+    public string ID { get; } = default!;
+
+    [DataField("parent")]
+    public string? Parent { get; } = default!;
```
Comment (mirrorcult) [inline Content.Server/Wires/WireLayout.cs:18]: shouldnt this use `IInheritingPrototype` and the `ParentDataField` attribute? serv3 already supports just doing inheritance for these

File: Content.Server/Wires/WiresSystem.cs [hunk 6]
```diff
@@ -0,0 +1,895 @@
...
+                }
+
+                data.Add(id, new WireLayout.WireData(d.Letter, d.Color, i));
+
+                wires.WiresList[i] = wireSet[id];
+            }
+
+            if (!wires.AlwaysRandomize)
+            {
+                AddLayout(wires.LayoutId, new WireLayout(data));
```
Comment (mirrorcult) [inline Content.Server/Wires/WiresSystem.cs:181]: build error here

File: Content.Server/Wires/WiresSystem.cs [hunk 7]
```diff
@@ -0,0 +1,895 @@
...
+        public TimedWireEvent OnFinish;
+
+        public ActiveWireAction(object identifier, float time, CancellationToken cancelToken, TimedWireEvent onFinish)
+        {
+            Id = identifier;
+            TimeLeft = time;
+            CancelToken = cancelToken;
+            OnFinish = onFinish;
+        }
+    }
```
Comment (mirrorcult) [inline Content.Server/Wires/WiresSystem.cs:357]: needs comments

File: Content.Shared/Wires/SharedWiresComponent.cs [hunk 8]
```diff
@@ -5,177 +5,173 @@
...
+        public WireLetter Letter;
+
+        public ClientWire(int id, bool isCut, WireColor color, WireLetter letter)
+        {
+            Id = id;
+            IsCut = isCut;
+            Letter = letter;
+            Color = color;
         }
     }
```
Comment (mirrorcult) [inline Content.Shared/Wires/SharedWiresComponent.cs:176]: comments?

**Review by mirrorcult — CHANGES_REQUESTED** [review]
generally good just think it needs more commenting

File: Content.Server/ParticleAccelerator/Components/ParticleAcceleratorControlBoxComponent.cs [hunk 9]
```diff
@@ -103,7 +106,7 @@ protected override void Startup()
         {
             base.Startup();
 
-            UpdateWireStatus();
+            // UpdateWireStatus();
```
[the remark this answers is not in the document]
Comment (vulppine) [reply Content.Server/ParticleAccelerator/Components/ParticleAcceleratorControlBoxComponent.cs:109]: this is so when somebody gets to refactoring the particle accelerator, they know what the wires did for this component (i could also just do the wires part myself, if needed)

File: Content.Server/Wires/WireLayout.cs (hunk 5 above)
Comment (vulppine) [reply Content.Server/Wires/WireLayout.cs:18]: oh nice, yeah i was wondering how to deal with this

File: Content.Server/Wires/WiresSystem.cs (hunk 6 above)
Comment (vulppine) [reply Content.Server/Wires/WiresSystem.cs:181]: whoops

File: Content.Server/Wires/WiresComponent.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    public string LayoutId { get; set; } = default!;
+
+    [ViewVariables]
+    public string? SerialNumber { get; set; }
+
+    [ViewVariables]
+    public int WireSeed { get; set; }
+
+    [ViewVariables]
+    public List<Wire> WiresList { get; set; } = new();
```
[the remark this answers is not in the document]
Comment (vulppine) [reply Content.Server/Wires/WiresComponent.cs:29]: honestly for what it's worth i really think that where should just be a direct index, IDs are just... indexes into the wire list? even in the original implementation it was just id++ attached to the wire itself so

[the remark this answers is not in the document]
Comment (vulppine) [reply Content.Server/Wires/WiresComponent.cs:29]: i feel like the wire.Id thing was done so that foreach could be used instead of a for loop

**Pull request merged**